sys/include/net/gcoap/dns.h

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright (C) 2021 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup net_gcoap_dns  DNS over CoAP client implementation
+ * @ingroup  net_gcoap
+ *
+ * @brief   A DNS over CoAP client prototype based on gCoAP.
+ *
+ * DNS over CoAP allows a node to use a CoAP server to resolve DNS request, following
+ * [draft-lenders-dns-over-coap](https://datatracker.ietf.org/doc/draft-lenders-dns-over-coap/).
+ *
+ * The `gcoap_dns` module does not replace the @ref sock_dns_query function when built, and is not
+ * used as a back-end to @ref netutils_get_ipv6 automatically. It does, however, provide a drop-in
+ * replacement for @ref sock_dns_query in its @ref gcoap_dns_query function.
+ *
+ * @{
+ *
+ * @file
+ * @brief   `gcoap_dns` definitions
+ *
+ * @author  Martine Lenders <m.lenders@fu-berlin.de>
+ */
+#ifndef NET_GCOAP_DNS_H
+#define NET_GCOAP_DNS_H
+
+#include <stdint.h>
+
+#include "mutex.h"
+#include "net/credman.h"
+#include "net/coap.h"
+#include "net/gcoap.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup    gcoap_dns_config    GCoAP DNS over CoAP client configuration
+ * @ingroup     config
+ * @{
+ */
+/**
+ * @brief   maximum length of the URI for the DNS server
+ */
+#ifndef CONFIG_GCOAP_DNS_SERVER_URI_LEN
+#define CONFIG_GCOAP_DNS_SERVER_URI_LEN     64U
+#endif
+
+/**
+ * @brief   maximum number of credentials that can be added with @ref gcoap_dns_cred_add()
+ */
+#ifndef CONFIG_GCOAP_DNS_CREDS_MAX
+#if IS_USED(MODULE_GCOAP_DTLS) || defined(DOXYGEN)
+#define CONFIG_GCOAP_DNS_CREDS_MAX          1U
+#else
+#define CONFIG_GCOAP_DNS_CREDS_MAX          0U
+#endif
+#endif
+
+/**
+ * @brief   size of the buffer used to build a CoAP request
+ */
+#ifndef CONFIG_GCOAP_DNS_PDU_BUF_SIZE
+#define CONFIG_GCOAP_DNS_PDU_BUF_SIZE       128U
+#endif
+
+/**
+ * @brief   block-size used for requests (both Block2 control and Block1)
+ *
+ * Should be less than @ref CONFIG_GCOAP_DNS_PDU_BUF_SIZE or
+ * @ref CONFIG_GCOAP_PDU_BUF_SIZE and must be a power
+ * of 2
+ */
+#ifndef CONFIG_GCOAP_DNS_BLOCK_SIZE
+#define CONFIG_GCOAP_DNS_BLOCK_SIZE         64U
+#endif
+/** @} */
+
+/**
+ * @brief   Query a domain name via CoAP synchronously
+ *
+ * @note    Only available with module `gcoap_dns_async` compiled in.
+ *
+ * @param[in] domain_name   A '\0'-terminated domain name. Must not be NULL.
+ * @param[out] addr_out     The resolved address. Must not be NULL.
```

Comment:
The typechecking nerd in me would love to have this a union pointer; if you think that excessive, it should at least be

```suggestion
 * @param[out] addr_out     The resolved address. Must not be NULL.
 *                          A typical target type is `* ipv6_addr_t` for AF_INET6.
```

(Solving this fully would probably involve changing things on the DNS side here, as `dns_msg_parse_reply` works that way).

Reply:
This is basically copied from the `sock_dns` API and IMHO the two APIs should align. Another typical target could be `sock_udp_ep_t`'s `addr` member, so I don't think we should specify here.